---
title: "Adversity-DNAm adolescence. Part 3: Top hits and biology"
output: html_document
---

*Setup*
```{r setup}
library(cowplot)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(readr)
library(reshape)
library(viridis)

setwd("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/")
load("Results/age15.results.top_2022-08-18.Rdata", verbose=T)
load("Results/age15.results.variable_2022-08-18.Rdata", verbose=T)

# write.table(age15.fdr.R2, file = "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Tables/Table1_topLoci_2022-12-05.txt",
#             sep="\t",quote=F, row.names=F, col.names=T)



```

*Figure 2 - top loci*
```{r Figure 2}
age15.results.variable$Adversity <- factor(age15.results.variable$Adversity, 
                                           levels = c("abuse","r_faminstability","Fscore",
                                                      "mompsych","nbhqual","oneadult",
                                                      "parcruelty"))
adv.labels <- c("Caregiver physical or emotional abuse", 
                "Sexual or physical abuse (by anyone)", 
                "Maternal psychopathology", 
                "One adult in the household", 
                "Family instability", 
                "Financial hardship", 
                "Neighborhood disadvantage")
names(adv.labels) <- levels(age15.fdr.R2$Adversity.2)

age15.results.variable$Adversity.2 <- factor(age15.results.variable$Adversity, 
                                           levels = c("parcruelty", "abuse", "mompsych",
                                                      "oneadult","r_faminstability", 
                                                      "Fscore","nbhqual"))
age15.results.variable$facetflag <- 
  factor(ifelse(age15.results.variable$Timing %in% c("accumulation", "recency"), 
                     "Additive", "Sensitive period"), 
              levels=c("Sensitive period", "Additive"))

age15.results.variable$timeclass <- ifelse(age15.results.variable$Timing =="very_early",
                                           "Very early\nchildhood", 
                                           ifelse(age15.results.variable$Timing =="early",
                                           "Early\nchildhood", 
                                           ifelse(age15.results.variable$Timing =="middle",
                                           "Middle\nchildhood", 
                                           ifelse(age15.results.variable$Timing =="late",
                                           "Late\nchildhood", 
                                           ifelse(age15.results.variable$Timing =="accumulation",
                                           "Accumulation", "Recency")))))
  
age15.results.variable$timeclass <- factor(age15.results.variable$timeclass, 
                                           levels = c("Very early\nchildhood", 
                                                "Early\nchildhood", "Middle\nchildhood",
                                                "Late\nchildhood",
                                                "Accumulation", "Recency"))


## FDR hits
age15.fdr <- age15.results.variable[age15.results.variable$FDR<=0.05,]

age15.fdr.2 <- rbind(age15.fdr, NA)

age15.fdr.2$timeclass[is.na(age15.fdr.2$timeclass)] <- "Middle\nchildhood"
age15.fdr.2$facetflag[is.na(age15.fdr.2$facetflag)] <- "Sensitive period"
age15.fdr.2 <- rbind(age15.fdr.2, NA)
age15.fdr.2$timeclass[is.na(age15.fdr.2$timeclass)] <- "Recency"
age15.fdr.2$facetflag[is.na(age15.fdr.2$facetflag)] <- "Additive"

p1 <- ggplot(age15.fdr.2, 
       aes(x=timeclass, fill =Adversity.2, col = Adversity.2))+
  geom_bar(stat='count', alpha=0.77)+
  theme_classic()+
  #facet_wrap(~Analysis)+
  # scale_fill_viridis("Adversity", discrete=T,
  #                     labels = c(adv.labels, ""), drop=F,
  #                     na.value="white") +
  scale_fill_discrete("Adversity",
                      labels = c(adv.labels, ""), drop=F,
                      guide = guide_legend(reverse=FALSE), na.value="white") +
  scale_color_manual("Adversity", values = rep("black",7),
                      labels = c(adv.labels, ""), drop=F,
                      guide = guide_legend(reverse=FALSE), na.value="white") + 
  facet_grid(~ facetflag, scales="free_x", space="free", drop=F) + 
  xlab("Life course model") + 
  ylab("Number of loci (FDR≤0.05)") +
  theme_bw() + 
  theme(text = element_text(family = "Arial"),
        axis.text.x = element_text(size=12, angle=45, hjust=1),
        axis.title.x = element_text(size=14),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=14),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14),
        strip.text = element_text(size=14),
        plot.title.position = "plot",
        plot.title = element_text(size=22, vjust = -2, face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        legend.position = "none")+
  geom_text(data=age15.fdr.2[!is.na(age15.fdr.2$CpG),],
            stat="count", aes(label =..count.., group=timeclass), vjust=-1, col ="black")+
  ylim(0,25)
p1

## FDR and R2 hits
age15.fdr.R2 <- age15.results.variable[age15.results.variable$R2 >=0.035,]
table(age15.fdr.R2$timeclass)

age15.fdr.R2.2 <- rbind(age15.fdr.R2, NA)
age15.fdr.R2.2$timeclass[is.na(age15.fdr.R2.2$timeclass)] <- "Middle\nchildhood"
age15.fdr.R2.2$facetflag[is.na(age15.fdr.R2.2$facetflag)] <- "Sensitive period"
# age15.fdr.R2.2 <- rbind(age15.fdr.R2.2, NA)
# age15.fdr.R2.2$timeclass[is.na(age15.fdr.R2.2$timeclass)] <- "Recency"
# age15.fdr.R2.2$facetflag[is.na(age15.fdr.R2.2$facetflag)] <- "Additive"

adv.labels.2 <- paste0(adv.labels, " (", c(3, 4, 1, 20, 1, 9, 3), ")")

p2 <- ggplot(age15.fdr.R2.2, 
       aes(x=timeclass, fill =Adversity.2, col = Adversity.2))+
  geom_bar(stat='count', alpha=0.77)+
  theme_classic()+
  #facet_wrap(~Analysis)+
  # scale_fill_viridis("Adversity (# CpGs at R2-threshold)", discrete=T,
  #                     labels = c(adv.labels.2, ""), drop=F,
  #                     na.value="white") + 
  scale_fill_discrete("Adversity",
                      labels = c(adv.labels, ""), drop=F,
                      guide = guide_legend(reverse=FALSE), na.value="white") +
  scale_color_manual("Adversity (# CpGs at R2-threshold)", values = rep("black",7),
                      labels = c(adv.labels, ""), drop=F,
                      guide = "none", na.value="white") + 
  facet_grid(~ facetflag, scales="free_x", space="free", drop=F) + 
  xlab("Life course model") + 
  ylab("Number of loci (R2≥0.035)") +
  theme_bw() + 
  theme(text = element_text(family = "Arial"),
        axis.text.x = element_text(size=12, angle=45, hjust=1),
        axis.title.x = element_text(size=14),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=14),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14),
        strip.text = element_text(size=14),
        plot.title.position = "plot",
        plot.title = element_text(size=22, vjust = -2, face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank())+
  geom_text(data=age15.fdr.R2.2[!is.na(age15.fdr.R2.2$CpG),],
            stat="count", aes(label =..count.., group=timeclass), vjust=-1, col ="black")+
  ylim(0,25)
p2
leg <- get_legend(p2)

gridExtra::grid.arrange(p1+ggtitle("A"),
                        p2+theme(legend.position ="none")+ggtitle("B"), 
                        leg, nrow=1)
g <- arrangeGrob(p1+ggtitle("A"),
           p2+theme(legend.position ="none")+ggtitle("B"), 
           leg, nrow=1)
ggsave(g, file = "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Submission 9 - Lancet Child Adolescent Health - Resubmission 2/Figures/Figure2_2023-04-17.eps", dpi = 300, width=14, height=6, device=cairo_ps)

# gridExtra::grid.arrange(p2+theme(legend.position ="none")+ggtitle("A"), 
#                         p1+ggtitle("B")+ylab("Number of loci (R2≥0.035 & FDR≤0.05"),
#                         leg, nrow=1)


## BONFERRONI
# ggplot(age15.results.variable[which(age15.results.variable$bonf<0.05),], 
#        aes(x=timeclass, fill =Adversity.2))+
#   geom_bar(stat='count',col="black")+
#   theme_classic()+
#   #facet_wrap(~Analysis)+
#   scale_fill_discrete("Adversity", 
#                       labels=adv.labels, drop=F,
#                       guide = guide_legend(reverse=FALSE)) + 
#   #facet_grid(~ facetflag, space="free") + 
#   xlab("Theoretical Model") + 
#   ylab("Number of significant loci (q<0.05)") +
#   theme_bw() + 
#   theme(text = element_text(family = "Arial"),
#                      axis.text.x = element_text(size=12, angle=45, hjust=1),
#                      axis.title.x = element_text(size=14),
#                      axis.text.y = element_text(size=12),
#                      axis.title.y = element_text(size=14),
#                      legend.text = element_text(size=12),
#                      legend.title = element_text(size=14),
#                      strip.text = element_text(size=14),
#                      panel.grid.major = element_blank(),
#                      panel.grid.minor = element_blank(),
#                      panel.background = element_blank())+
#   scale_x_discrete(drop=F)

rm(p1, p2, leg)

save(age15.fdr.R2, age15.fdr, file = "Results/age15.results.top_2022-08-24.Rdata")

```

*Figure S19 - deprivation vs threat*
```{r Figure S19}

age15.fdr.2$Deprivation <- ifelse(age15.fdr.2$Adversity %in% c("abuse","parcruelty"),"Threat",
                                  ifelse(age15.fdr.2$Adversity %in% c("oneadult","mompsych","nbhqual",
                                                                      "r_faminstability","Fscore"),
                                         "Deprivation",NA))
age15.fdr.2$Deprivation <- factor(age15.fdr.2$Deprivation)
age15.fdr.R2.2$Deprivation <- ifelse(age15.fdr.R2.2$Adversity %in% c("abuse","parcruelty"),"Threat",
                                     ifelse(age15.fdr.R2.2$Adversity %in%
                                              c("oneadult","mompsych","nbhqual",
                                                "r_faminstability","Fscore"), "Deprivation",NA))
age15.fdr.R2.2$Deprivation <- factor(age15.fdr.R2.2$Deprivation)

p1 <- ggplot(age15.fdr.2, 
       aes(x=timeclass, fill =Deprivation, col = Deprivation))+
  geom_bar(stat='count', alpha=0.77, position = position_dodge2(preserve="single"))+
  theme_classic()+
  #facet_wrap(~Analysis)+
  scale_fill_manual("", 
                    labels = c("Deprivation","Threat",""), drop=F,
                    values = c(viridis(n=2, end = 0.8)), na.value="white")+
  scale_color_manual("", 
                    labels = c("Deprivation","Threat",""), drop=F,
                    values = rep("black",2), na.value="white") + 
  facet_grid(~ facetflag, scales="free_x", space="free", drop=F) + 
  xlab("Life course model") + 
  ylab("Number of loci (FDR≤0.05)") +
  theme_bw() + 
  theme(text = element_text(family = "Arial"),
        axis.text.x = element_text(size=12, angle=45, hjust=1),
        axis.title.x = element_text(size=14),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=14),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14),
        strip.text = element_text(size=14),
        plot.title.position = "plot",
        plot.title = element_text(size=22, vjust = -2, face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        legend.position = "none")+
  geom_text(data=age15.fdr.2[!is.na(age15.fdr.2$CpG),],
            stat="count", aes(label =..count.., group=Deprivation, y =..count..), 
            position = position_dodge(w=0.9), vjust=-1, col ="black")+
  ylim(0,20)
p1

p2 <- ggplot(age15.fdr.R2.2, 
       aes(x=timeclass, fill =Deprivation, col = Deprivation))+
  geom_bar(stat='count', alpha=0.77, position = position_dodge2(preserve = "single"))+
  theme_classic()+
  #facet_wrap(~Analysis)+
  scale_fill_manual("", 
                    labels = c("Deprivation","Threat",""), drop=F,
                    values = c(viridis(n=2, end = 0.8)), na.value="white")+
  scale_color_manual("", 
                    labels = c("Deprivation","Threat",""), drop=F,
                    values = rep("black",2), na.value="white") + 
  facet_grid(~ facetflag, scales="free_x", space="free", drop=F) + 
  xlab("Life course model") + 
  ylab("Number of loci (R2≥0.035)") +
  theme_bw() + 
  theme(text = element_text(family = "Arial"),
        axis.text.x = element_text(size=12, angle=45, hjust=1),
        axis.title.x = element_text(size=14),
        axis.text.y = element_text(size=12),
        axis.title.y = element_text(size=14),
        legend.text = element_text(size=12),
        legend.title = element_text(size=14),
        strip.text = element_text(size=14),
        plot.title.position = "plot",
        plot.title = element_text(size=22, vjust = -2, face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank())+
  geom_text(data=age15.fdr.R2.2[!is.na(age15.fdr.R2.2$CpG),],
            stat="count", aes(label =..count.., group=Deprivation, y =..count..), 
            position = position_dodge(w=0.9), vjust=-1, col ="black")+
  ylim(0,20)
p2
leg <- cowplot::get_legend(p2)

gridExtra::grid.arrange(p1+ggtitle("A"),
                        p2+theme(legend.position ="none")+ggtitle("B"), 
                        leg, nrow=1)

my.chisq.test <- function(x, y){
  p.1 <- unname(table(x))/sum(unname(table(x)))
  p.2 <- unname(table(y))/sum(unname(table(y)))
  
  ch <- chisq.test(unname(table(x)), p=p.2)
  
  return(c(FDR=p.1, fData=p.2, ch$stat, P=ch$p.val))
}

x <- age15.fdr.R2.2$timeclass[age15.fdr.R2.2$Deprivation =="Deprivation"]
x <- unname(table(x))/sum(unname(table(x)))
y <- age15.fdr.R2.2$timeclass[age15.fdr.R2.2$Deprivation =="Threat"]
y <- unname(table(y))/sum(unname(table(y)))
chisq.test(x,y)
# X-squared = 7, df = 6, p-value = 0.3208

rm(age15.fdr.2, age15.fdr.R2.2, p1, p2, leg, x,y)
```


*Regulatory elements - Figure S2*
```{r figure S2}
#prep data 
load("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/featureData.Rdata", 
     verbose=T)
featureData$CHR37 <- as.character(featureData$CHR37)
featureData$COORDINATE_37 <- as.character(featureData$COORDINATE_37)
fData <- featureData[featureData$TargetID %in% age15.results.variable$CpG, ]
dim(fData) #302581 CpGs
rm(featureData)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data(Other)
data(Islands.UCSC)
data(Locations)

#FDR results
fdr2 <- age15.fdr %>% dplyr::select(CpG)
anno.fdr <- Other[fdr2$CpG, ]
cgi.fdr <- Islands.UCSC[fdr2$CpG, ]
cgi.fdr$CpG <- rownames(cgi.fdr)

fdr2 <- merge(fdr2, anno.fdr[, c(10, 13, 14)], by.x="CpG", by.y="row.names")
fdr2 <- merge(fdr2, cgi.fdr, by="CpG")
colnames(fdr2)

fdr2$Enhancer2 <- ifelse(fdr2$Enhancer == TRUE, 1, 0)
fdr2$Promoter.Assoc <- ifelse(fdr2$Regulatory_Feature_Group %in% 
                                c("Promoter_Associated", "Promoter_Associated_Cell_type_specific"), 1, 0)
anno.autosome <- Other[fData$TargetID, ]
anno.autosome <- anno.autosome[!is.na(rownames(anno.autosome)), ]
apply(anno.autosome[, c(10, 13, 14)], 2, table)

cgi.auto <- Islands.UCSC[fData$TargetID, ]
cgi.auto <- cgi.auto[!is.na(rownames(cgi.auto)), ]
table(cgi.auto[, 2])

#BACKGROUND DATA 
fData2 <- merge(fData, anno.autosome[, c(10, 13, 14)], by.x="TargetID", by.y="row.names")
fData2 <- as.data.frame(fData2@listData)
fData2 <- merge(fData2, cgi.auto, by.x="TargetID", by.y="row.names")
fData2 <- as.data.frame(fData2@listData)

fData2$Enhancer2 <- ifelse(fData2$Enhancer == TRUE, 1, 0)
fData2$Promoter.Assoc <- ifelse(fData2$Regulatory_Feature_Group %in% 
                                  c("Promoter_Associated", "Promoter_Associated_Cell_type_specific"), 1, 0)
apply(fData2[, 22:23], 2, table)


#TESTING ENRICHMENT
my.chisq.test <- function(x, y){
  p.1 <- unname(table(x))/sum(unname(table(x)))
  p.2 <- unname(table(y))/sum(unname(table(y)))
  
  ch <- chisq.test(unname(table(x)), p=p.2)
  
  return(c(FDR=p.1, fData=p.2, ch$stat, P=ch$p.val))
}

ch1  <- my.chisq.test(x=fdr2$Enhancer2, y=fData2$Enhancer2)
ch1
#       FDR1       FDR2     fData1     fData2  X-squared          P 
# 0.54545455 0.45454545 0.74234007 0.25765993 4.45861994 0.03472554 

fdr2$Promoter.Assoc <- factor(fdr2$Promoter.Assoc, levels =c (0,1))
ch2 <- my.chisq.test(x=fdr2$Promoter.Assoc, y=fData2$Promoter.Assoc)
ch2
#       FDR1       FDR2     fData1     fData2  X-squared          P 
# 1.00000000 0.00000000 0.91960169 0.08039831 1.92340092 0.16548223 
fdr2$DHS <- factor(fdr2$DHS, levels = c("","TRUE"))
ch3<- my.chisq.test(x = fdr2$DHS, y= fData2$DHS)
ch3
#      FDR1      FDR2    fData1    fData2 X-squared         P 
# 1.0000000 0.0000000 0.8946563 0.1053437 2.5904487 0.1075099 
fdr2$Relation_to_Island <- factor(fdr2$Relation_to_Island, 
                                  levels = levels(fData2$Relation_to_Island))
ch4 <- my.chisq.test(x=fdr2$Relation_to_Island, y=fData2$Relation_to_Island)
ch4
 #      FDR1        FDR2        FDR3        FDR4        FDR5        FDR6      fData1      fData2      fData3 
 # 0.00000000  0.04545455  0.04545455  0.81818182  0.00000000  0.09090909  0.17379148  0.06453479  0.14274194 
 #     fData4      fData5      fData6   X-squared           P 
 # 0.45190544  0.05786550  0.10916085 13.27765871  0.02091119


#R2 HITS
fdr.r2 <- age15.fdr.R2 %>% dplyr::select(CpG)
anno.fdr <- Other[fdr.r2$CpG, ]
cgi.fdr <- Islands.UCSC[fdr.r2$CpG, ]
cgi.fdr$CpG <- rownames(cgi.fdr)

fdr.r2 <- merge(fdr.r2, anno.fdr[, c(10, 13, 14)], by.x="CpG", by.y="row.names")
fdr.r2 <- merge(fdr.r2, cgi.fdr, by="CpG")
colnames(fdr.r2)

fdr.r2$Enhancer2 <- ifelse(fdr.r2$Enhancer == TRUE, 1, 0)
fdr.r2$Promoter.Assoc <- ifelse(fdr.r2$Regulatory_Feature_Group %in% 
                                c("Promoter_Associated", "Promoter_Associated_Cell_type_specific"), 1, 0)
anno.autosome <- Other[fData$TargetID, ]
anno.autosome <- anno.autosome[!is.na(rownames(anno.autosome)), ]
apply(anno.autosome[, c(10, 13, 14)], 2, table)

cgi.auto <- Islands.UCSC[fData$TargetID, ]
cgi.auto <- cgi.auto[!is.na(rownames(cgi.auto)), ]
table(cgi.auto[, 2])

ch1.r  <- my.chisq.test(x=fdr.r2$Enhancer2, y=fData2$Enhancer2)
ch1.r
#        FDR1        FDR2      fData1      fData2   X-squared           P 
# 0.560975610 0.439024390 0.742340068 0.257659932 7.050800557 0.007922983 

fdr.r2$Promoter.Assoc <- factor(fdr.r2$Promoter.Assoc, levels =c (0,1))
ch2.r <- my.chisq.test(x=fdr.r2$Promoter.Assoc, y=fData2$Promoter.Assoc)
ch2.r
#       FDR1       FDR2     fData1     fData2  X-squared          P 
# 0.95121951 0.04878049 0.91960169 0.08039831 0.55437164 0.45653693 
fdr.r2$DHS <- factor(fdr.r2$DHS, levels = c("","TRUE"))
ch3.r <- my.chisq.test(x = fdr.r2$DHS, y= fData2$DHS)
ch3.r
#       FDR1       FDR2     fData1     fData2  X-squared          P 
# 0.97560976 0.02439024 0.89465631 0.10534369 2.85095134 0.09131988
fdr.r2$Relation_to_Island <- factor(fdr.r2$Relation_to_Island, 
                                  levels = levels(fData2$Relation_to_Island))
ch4.r <- my.chisq.test(x=fdr.r2$Relation_to_Island, y=fData2$Relation_to_Island)
ch4.r
 #       FDR1        FDR2        FDR3        FDR4        FDR5        FDR6      fData1      fData2      fData3 
 # 0.02439024  0.04878049  0.07317073  0.70731707  0.02439024  0.12195122  0.17379148  0.06453479  0.14274194 
 #     fData4      fData5      fData6   X-squared           P 
 # 0.45190544  0.05786550  0.10916085 13.58773233  0.01845154


### PLOTS - FIGURE S2 ### 
d1 <- cbind.data.frame(probeset=factor(rep(c("FDR-significant sites", 
                                             "R2-threshold sites",
                                             "All sites tested"), 2), 
                                       levels=c("FDR-significant sites", 
                                                "R2-threshold sites",
                                                "All sites tested")), 
                       annot=factor(c(rep("Promoter", 3), rep("Enhancer", 3)),
                                    levels=c("Promoter", "Enhancer")), 
                       prop = c(ch2[2], ch2.r[c(2, 4)], ch1[2], ch1.r[c(2,4)]))
d1
p1 <- ggplot(d1, aes(x=annot, y=prop, group=probeset)) + 
  geom_bar(stat="identity", aes(fill=probeset), position = "dodge", width=0.7, 
           col = 'black', alpha =0.77) + 
  scale_fill_manual(values=c("steelblue4", "steelblue1", "grey")) + 
  labs(x=NULL, y="Proportion") + 
  theme_bw() + 
  ggtitle("A")+
  theme(plot.title.position = "plot",
        plot.title = element_text(face="bold"))+
  theme(legend.title = element_blank(), 
        panel.grid.major.x = element_blank(), 
        text = element_text(family = "Gill Sans", size = 22))
p1
# next:
fdr = table(fdr2$Relation_to_Island)/sum(unname(table(fdr2$Relation_to_Island)))
r2 = table(fdr.r2$Relation_to_Island)/sum(unname(table(fdr.r2$Relation_to_Island)))
rest = table(fData2$Relation_to_Island)/sum(unname(table(fData2$Relation_to_Island)))

d2 <- cbind.data.frame(probeset=factor(c(rep("FDR-significant sites", 4),
                                         rep("R2-threshold sites", 6),
                                         rep("All sites tested", 6)), 
                                       levels=c("FDR-significant sites", 
                                                "R2-threshold sites",
                                                "All sites tested")), 
                       annot=factor(c(names(fdr), names(r2), names(rest)),
                                    levels=c("OpenSea", "N_Shelf", "N_Shore", 
                                             "Island", "S_Shore", "S_Shelf"),
                                    labels=c("Open Sea", "N. Shelf", "N. Shore", 
                                             "Island", "S. Shore", "S. Shelf")), 
                       prop=c(unname(fdr), unname(r2), unname(rest)))
rm(fdr, rest)

p2 <- ggplot(d2, aes(x=annot, y=prop, group=probeset)) + 
  geom_bar(stat="identity", aes(fill=probeset), position = "dodge", width=0.7, 
           col ='black', alpha = 0.77) + 
  scale_fill_manual(values=c("steelblue4", "steelblue1", "grey")) + 
  labs(x=NULL, y="Proportion") + 
  theme_bw() + 
  theme(plot.title.position = "plot",
        plot.title = element_text(face="bold"))+
  ggtitle("B") + 
  theme(legend.title = element_blank(), 
        panel.grid.major.x = element_blank(), 
        text = element_text(family = "Gill Sans", size = 22))
p2

grid.arrange(p1, p2, ncol = 1)
  

rm(anno.fdr, cgi.fdr, d1, d2, 
   fData2, my.chisq.test,p1, p2, ch1, ch2, ch3, ch4)
rm(fData, anno.autosome, cgi.auto, Other, Islands.UCSC)
rm(ch1.r, ch2.r, ch3.r, ch4.r, Locations, r2, fdr, rest)

```

*Compare blood and brain for top loci - Figure S3*
```{r Figure S3}
#had to grab these manually from https://epigenetics.essex.ac.uk/bloodbrain/
corrs <- gdata::read.xls("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Tables/brain_blood_correlation.age15.hits_2022-08-19.xlsx")
head(corrs)
dim(corrs)
sum(corrs$Probe %in% age15.fdr$CpG)
sum(corrs$Probe %in% age15.fdr.R2$CpG)

#FDR
corrs.fdr <- corrs[corrs$FDR==1,]
summary(corrs.fdr[,4:7])
colMeans(corrs.fdr[,4:7]) 
#        PFC         EC        STG        CER 
# 0.04964545 0.06014182 0.04724682 0.05671364 
colMeans(abs(corrs.fdr[rowMeans(corrs.fdr[,4:7])>0, 4:7]))
#       PFC        EC       STG       CER 
# 0.1431014 0.1476979 0.1445988 0.1470030 
sum(rowMeans(corrs.fdr[,4:7])>0) #17/22

colMeans((corrs.fdr[rowMeans(corrs.fdr[,4:7])<0, 4:7]))
 #     PFC       EC      STG      CER 
 # 0.02306 -0.02700 -0.02894 -0.06730


#R2
summary(corrs[,4:7])
colMeans(corrs[,4:7]) 
#        PFC         EC        STG        CER 
# 0.10807205 0.11272732 0.09268683 0.08654146 
colMeans(abs(corrs[rowMeans(corrs[,4:7])>0, 4:7]))
#       PFC        EC       STG       CER 
# 0.1960500 0.1962689 0.1878761 0.1761750
sum(rowMeans(corrs[,4:7])>0) #28/41

colMeans((corrs[rowMeans(corrs[,4:7])<0, 4:7]))
#          PFC           EC          STG          CER 
# -0.003695846 -0.010080769 -0.020551538 -0.084592308

corrs.melt <- melt(rbind(data.frame(corrs.fdr[,-c(2:3)], Type = "FDR-significant"), 
                         data.frame(corrs[,-c(2:3)], Type = "R2-threshold")))

ggplot(corrs.melt, aes(x= variable, y = value))+
  geom_violin()+
  geom_boxplot(width= 0.2, outlier.size = -1, fill='grey')+
  geom_jitter(width = 0.05,height = 0, alpha = 0.7, aes(col = variable))+
  theme_classic()+
  geom_hline(yintercept =0, linetype = 3)+
  facet_wrap(~Type)+
  scale_color_viridis(option = 'magma', discrete=T, "Brain Region", end = 0.7)+
  xlab("Brain region")+
  ylab("Brain/Blood DNAm correlation")+
  ylim(-0.5,1)+
  theme(text = element_text(family = "Arial"),
        axis.text = element_text(size =12), 
        axis.title = element_text(size = 14), 
        strip.text = element_text(size = 12))
    
rm(corrs, corrs.fdr, corrs.melt)
```

*Annotated table*
```{r}
fdr.results <- age15.fdr.R2

library(FDb.InfiniumMethylation.hg19)
library(tibble)
load("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/featureData.Rdata", verbose=T)
hm450 <- get450k()
probes <- hm450[fdr.results$CpG]
x <- getNearestGene(probes)
summary(x$distance)
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0       0.0       0.0   10947.1     389.8 1095065.0 
x <- x %>% tibble::rownames_to_column("CpG")

fdr.annot <- left_join(fdr.results,x, by =  "CpG")

# UCSC:
fData <- featureData[row.names(featureData) %in% fdr.results$CpG, ]
rm(featureData)
fData <- fData %>% tibble::rownames_to_column("CpG")
fdr.annot <- left_join(fdr.annot, fData, by = "CpG")

## clean up the format 
fdr.annot <- fdr.annot %>% 
  dplyr::select(one_of(c(colnames(fdr.results), "CHR37", "COORDINATE_37", "nearestGeneSymbol", "distance", "RELATION_TO_UCSC_CPG_ISLAND"))) %>% 
  dplyr::rename(Chr = CHR37, Coordinate = COORDINATE_37, NearestGene = nearestGeneSymbol, Relation_to_CGI = RELATION_TO_UCSC_CPG_ISLAND) ## lookin' good 

fdr.annot$Chr <- unlist(fdr.annot$Chr)
fdr.annot$Coordinate <- unlist(fdr.annot$Coordinate)
# remove those further than 300 bp 
fdr.annot <- fdr.annot %>% 
  mutate(NearestGene = ifelse(distance > 300000, NA, NearestGene)) %>% 
  mutate(distance = ifelse(distance > 300000, NA, distance))
lapply(fdr.annot, class)
sum(is.na(fdr.annot$NearestGene)) # one did not have a gene annotated; that's okay 
head(fdr.annot)

## add whether it's a noisy probe identified by Naeem et al. 
naeem <- read_csv("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/FilteringCpGs_Naeem.csv") %>% filter(`Flag(discard/keep)` == "discard")
## included the following 2 files: 48639-non-specific-probes-Illumina450k.xlsx, 48640-polymorphic-CpGs-Illumina450k.xlsx
chen <- read_csv("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/chen-noisyprobes/chen2013_noisyprobes.csv", 
                 col_names = "Probe")
chen <- unique(chen) 
dim(chen)# n = 91698 

fdr.annot <- fdr.annot %>% 
  mutate(Noisy.inNaeem = ifelse(CpG %in% naeem$probe, 1, 0),
         Noisy.inChen = ifelse(CpG %in% chen$Probe, 1, 0))
rm(naeem, chen)

save(fdr.annot, 
     file = "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Tables/aim3.hits.annot_2022-08-19.Rdata")

## additional functional annotation 
# source("https://bioconductor.org/biocLite.R")
# biocLite("IlluminaHumanMethylation450kanno.ilmn12.hg19")
## Enhancer or promoter
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
data(IlluminaHumanMethylation450kanno.ilmn12.hg19)
annotation.table = getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
annotation.table <- annotation.table[row.names(annotation.table) %in% fdr.annot$CpG,]

annotation.table <- as.data.frame(annotation.table) %>% tibble::rownames_to_column("CpG")
colnames(annotation.table)
## it seems like the relation to island annotated here might be more accurate 
fdr.annot2 <- left_join(fdr.annot, annotation.table, by = "CpG")

table(fdr.annot2$Relation_to_Island)

# promoter: Promoter_Associated or Promoter_Associated_Cell_type_specific
fdr.annot2 <- fdr.annot2 %>% 
  mutate(Relation_to_CGI = Relation_to_Island,
         Enhancer = ifelse(Enhancer == TRUE, 1, 0),
         Promoter = ifelse(grepl("^Promoter", Regulatory_Feature_Group), 1, 0))

# select columns we'll need in the table 
fdr.annot2 <- fdr.annot2 %>% 
  dplyr::select(CpG, Adversity, Model, Timing, Beta, R2, P.value, FDR, bonf,
                Chr, Coordinate, NearestGene, distance, 
                Noisy.inNaeem, Noisy.inChen, Relation_to_CGI, Enhancer, Promoter)
fdr.annot2
# export a list of FDR CpGs for eFORGE lookups 
write.table(fdr.annot2$CpG[fdr.annot2$FDR<0.05], 
            "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/eForge/aim3_FDR_CpGs_eForge_2022-08-19.txt", 
            row.names = F, quote = FALSE)

write.table(fdr.annot2$CpG, 
            "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/eForge/aim3_R2_CpGs_eForge_2022-08-19.txt", 
            row.names = F, quote = FALSE)


```

*eForge*
```{r}
# columns O:T
# eFORGE: https://eforge.altiusinstitute.org/
# paste in the list of FDR significant CpGs
# Consolidated Roadmap Epigenomics - DHS
# Consolidated Roadmap Epigenomics - All H3
# Consolidated Roadmap Epigenomics - All 15 chromatin marks
# ENCODE - DHS
# followng Tom's code: Tom_Rcode_20181125_eFORGE func annot.R
saveDir <- "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/eForge/R2/"
annots <- cbind.data.frame(probeset = c(rep("Unnamed", 4)), 
                           annotset = c("erc2-DHS", "erc2-H3-all", 
                                        "encode","erc2-chromatin15state-all"))

# function to test whether there were any enrichments in any cell type were significant 
# after multiple corrections
# (actually pulling out what is most significant enrichment across each dataset)
check.sig <- function(x){
  # load data
  d <- read.delim(file=file.path(saveDir, 
                                 paste0(annots[x, 1], 
                                        ".450k.", annots[x, 2], 
                                        ".chart.tsv")),
                  sep="\t", header=T, stringsAsFactors = F)
  # subset to signficant associations
  d.sig <- d[d$Qvalue < 0.05, ]
  # pull out most significant association
  best <- d[which.min(d$Pvalue), ]
  # return all sig and and most sig
  return(list(as.character(annots[x, 2]), d.sig$Tissue, d.sig$Cell, 
              best$Tissue, best$Cell, best$Pvalue))
}
q <- sapply(c(1:3), check.sig) #not including all chromatin here 
q 
#nothing at q<0.05; strongest association in E028 Breast variant Human Mammary Epithelial Cells (p= 0.0107 )


# At some point we decided to report whether each CpG was within 1000bp of each of these regulatory marks
# in any of the cell/tissue types, and whether each CpG was within 1000bp of each of these marks in any 
# neuronal cells/brain tissue, and here is that function.
func <- function(x){
  d <- read.delim(file=file.path(saveDir, 
                                 paste0(annots[x, 1], 
                                        ".450k.", annots[x, 2], 
                                        ".chart.tsv")),
                  sep="\t", header=T, stringsAsFactors = F)
  
  d.anycell <- unique(unlist(strsplit(d$Probe, "[,]")))
  
  d.brain <- d[d$Cell %in% 
                 c("E007 H1 Derived Neuronal Progenitor Cultured Cells", 
                   "E081 Fetal Brain Male", 
                   "E082 Fetal Brain Female"), ]
  d.braincell <- unique(unlist(strsplit(d.brain$Probe, "[,]")))
  
  return(list(d.anycell, d.braincell))
}
r <- lapply(c(1:2), func)

# Cells named differently in ENCODE, so YIWEN made a separate function:
func2 <- function(x){
  d <- read.delim(file=file.path(saveDir, 
                                 paste0(annots[x, 1], 
                                        ".450k.", annots[x, 2], 
                                        ".chart.tsv")),
                  sep="\t", header=T, stringsAsFactors = F)
  
  d.anycell <- unique(unlist(strsplit(d$Probe, "[,]")))
  
  d.brain <- d[d$Tissue %in% 
                 c("Brain", "Brain hippocampus", "Cerebellar"), ]
  d.braincell <- unique(unlist(strsplit(d.brain$Probe, "[,]")))
  
  return(list(d.anycell, d.braincell))
}
# ENCODE result
s <- lapply(3, func2)

## add the information to FDR table
fdr.annot2 <- fdr.annot2 %>% 
  mutate(ERC_DHS_any.cell = ifelse(CpG %in% r[[1]][[1]], 1, 0),
         ERC_DHS_brain.cell = ifelse(CpG %in% r[[1]][[2]], 1, 0),
         ERC_H3.all_any.cell = ifelse(CpG %in% r[[2]][[1]], 1, 0),
         ERC_H3.all_brain.cell = ifelse(CpG %in% r[[2]][[2]], 1, 0),
         ENCODE_DHS_any.cell = ifelse(CpG %in% s[[1]][[1]], 1, 0),
         ENCODE_DHS_brain.cell = ifelse(CpG %in% s[[1]][[2]], 1, 0)
         )

apply(fdr.annot2[, 17:22], 2, table) # sanity check


save(fdr.annot2, file = "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/aim3_annot_table_2022-08-19.Rdata")
write.table(fdr.annot2, 
            file ="~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Tables/aim3_annot_table_2022-08-19.txt",
            sep="\t", quote=F, row.names = F, col.names=T)



#checking results R2
saveDir <- "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/eForge/R2/"
annots <- cbind.data.frame(probeset = c(rep("Unnamed", 4)), 
                           annotset = c("erc2-DHS", "erc2-H3-all", 
                                        "encode","erc2-chromatin15state-all"))

# function to test whether there were any enrichments in any cell type were significant 
# after multiple corrections
# (actually pulling out what is most significant enrichment across each dataset)
q <- sapply(c(1:3), check.sig) #not including all chromatin here 
q 
#nothing at q<0.05; strongest association in E028 Breast variant Human Mammary Epithelial Cells (p= 0.0107 )

#checking results FDR
saveDir <- "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/eForge/FDR/"
annots <- cbind.data.frame(probeset = c(rep("Unnamed", 4)), 
                           annotset = c("erc2-DHS", "erc2-H3-all", 
                                        "encode","erc2-chromatin15state-all"))

# function to test whether there were any enrichments in any cell type were significant 
# after multiple corrections
# (actually pulling out what is most significant enrichment across each dataset)
q <- sapply(c(1:3), check.sig) #not including all chromatin here 
q 
#nothing at q<0.05; strongest association in SKMC (p= 0.0153  ) and E008 H9 Cells (p = 0.0161)


rm(annots, check.sig, func, func2, q, r,s, saveDir, fdr.annot2)

rm(fData, fdr.annot, fdr.annot2, fdr.results, hm450, hm450.controls, IlluminaHumanMethylation450kanno.ilmn12.hg19, probes, x, seqinfo.hg19, annotation.table)

```


*Intolerance to LOF*
```{r}
exac <- read.delim("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/forweb_cleaned_exac_r03_march16_z_data_pLI.txt",
                   sep="\t", header=T)
hist(exac$pLI)
exac$gene <- as.character(levels(exac$gene))[exac$gene]

load("~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/featureData.Rdata", verbose=T)
featureData$CHR37 <- as.character(featureData$CHR37)
featureData$COORDINATE_37 <- as.character(featureData$COORDINATE_37)
fData <- featureData[featureData$TargetID %in% age15.results.variable$CpG, ]
dim(fData) #302581 CpGs
rm(featureData)

# annotate to nearestGene:
library(FDb.InfiniumMethylation.hg19)
hm450 <- get450k()
probes <- hm450[unique(fData$TargetID)]
x <- getNearestGene(probes)
# since some are beyond 300 kb, need to remove annotations:
x <- x[x$distance <= 300000, ]
length(unique(x$nearestGeneSymbol))
#[1] 20219

exac <- exac[exac$gene %in% x$nearestGeneSymbol, ]
nrow(exac)
#[1] 16114
rm(fData, probes, hm450, x, hm450.controls, seqinfo.hg19)

load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/aim3_annot_table_2022-08-19.Rdata", verbose=T)

rm(fdr.annot)
exac$fdr.gene.flag <- ifelse(exac$gene %in% fdr.annot2$NearestGene[fdr.annot2$FDR<0.05], 1, 0)
exac$r2.gene.flag <- ifelse(exac$gene %in% fdr.annot2$NearestGene, 1, 0)
unique(fdr.annot2$NearestGene)
table(exac$fdr.gene.flag) #15
table(exac$r2.gene.flag) #30
t.test(pLI ~ fdr.gene.flag, data=exac) #p-value = 0.3644 for FDR
t.test(pLI ~ r2.gene.flag, data=exac) #p-value = 0.8589 for R2
sum(fdr.annot2$NearestGene %in% exac$gene[exac$pLI>0.9 & exac$fdr.gene.flag ==1]) #3/15 genes 
sum(fdr.annot2$NearestGene %in% exac$gene[exac$pLI>0.9 & exac$r2.gene.flag ==1]) #7/30 genes 
fdr.annot2[fdr.annot2$NearestGene %in% exac$gene[exac$pLI>0.9 & exac$r2.gene.flag ==1],]

fdr.ex <- exac[exac$fdr.gene.flag ==1, c("gene","pLI") ]
r2.ex <- exac[exac$r2.gene.flag ==1, c("gene","pLI") ]

fdr.annot2$pLI <- r2.ex$pLI[match(fdr.annot2$NearestGene, r2.ex$gene)]

write.table(fdr.annot2, file ="~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Tables/aim3_annot_table_full_2022-08-31.txt",
            sep="\t", quote=F, row.names = F, col.names=T)

# probably better to test with permutations:
# following (http://faculty.washington.edu/kenrice/sisg/SISG-08-06.pdf)
# function to get mean for single permutation test:
one.test <- function(x,y) {
  xstar<-sample(x)
  mean(y[xstar==1])-mean(y[xstar==0])
}
one.test(exac$fdr.gene.flag, exac$pLI)
# [1] -0.0647337
# repeat 10k times:
many.perms <- replicate(10000, one.test(exac$fdr.gene.flag, exac$pLI))
# how many means were more extreme than the one we observed?
(group.diff <- mean(exac$pLI[exac$fdr.gene.flag==1]) -
    mean(exac$pLI[exac$fdr.gene.flag==0]))
mean(abs(many.perms) > abs(group.diff))
#[1] 0.27 for FDR

(group.diff <- mean(exac$pLI[exac$r2.gene.flag==1]) -
    mean(exac$pLI[exac$r2.gene.flag==0]))
many.perms <- replicate(10000, one.test(exac$r2.gene.flag, exac$pLI))
mean(abs(many.perms) > abs(group.diff))
#[1] 0.51 for R2

#plot:
hist(many.perms, xlim=c(-abs(group.diff), abs(group.diff)))
abline(v=group.diff, lwd=2, col="purple")

exax.melt <- rbind(data.frame(exac, subset= "All sites tested"),
                   data.frame(exac[exac$fdr.gene.flag ==1,], subset="FDR-significant sites"),
                   data.frame(exac[exac$r2.gene.flag ==1,], subset= "R2-threshold sites"))
exax.melt$subset <- factor(exax.melt$subset, 
                           levels = c("FDR-significant sites","R2-threshold sites","All sites tested"))

ggplot(exax.melt, aes(x=subset, y=pLI, group=subset)) + 
  geom_violin(aes(fill=subset), show.legend = F, alpha=0.77) + 
  scale_fill_manual(values=c("steelblue4", "steelblue1", "grey")) + 
  stat_summary(fun=mean, geom="point", size=3, colour="black") + 
  geom_jitter(data = exax.melt[exax.melt$subset == "FDR-significant sites",], 
               width = 0.05, shape = 3, size =3)+
  geom_jitter(data = exax.melt[exax.melt$subset == "R2-threshold sites",], 
               width = 0.05, shape = 3, size =3)+
  labs(x=NULL, y="Intolerance to loss-of-function (pLI)") + 
  theme_bw() + 
  theme(panel.grid.major.x = element_blank(), 
        text = element_text(family = "Arial", size = 22))

rm(exac, many.perms, one.test, group.diff, fdr.annot2, exax.melt, r2.ex, fdr.ex)
```


*Gene ontology*
```{r Figure S4}
load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/aim3_annot_table_2022-08-19.Rdata")
dim(fdr.annot2)
r2.genes <- c(na.omit(unique(fdr.annot2$NearestGene))) 
length(r2.genes)# 39 unique genes
fdr.genes <- c(na.omit(unique(fdr.annot2$NearestGene[fdr.annot2$FDR<0.05]))) 
length(fdr.genes)# 21 unique genes

write.table(fdr.genes, "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/DAVID/age15.FDRgenesforDAVID.txt", 
            quote = FALSE, sep = "\n", row.names = F)

write.table(r2.genes, "~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/DAVID/age15.R2genesforDAVID.txt", 
            quote = FALSE, sep = "\n", row.names = F)


# export background list of genes for DAVID:
background <- names(y)
# need EntrezIDs:
library(org.Hs.eg.db)
library(reshape2)
xx <- as.list(org.Hs.egALIAS2EG)
xx <- xx[!is.na(xx)]
xx <- xx[background]
xx <- melt(xx)
names(xx) <- c("EntrezID", "Symbol")
xx$EntrezID <- as.character(levels(xx$EntrezID))[xx$EntrezID]
dim(xx)
#[1] 23593     2
length(unique(xx$Symbol))
#[1] 21862

xx <- xx[!duplicated(xx$EntrezID), ]
dim(xx)
#[1] 22606     2
length(unique(xx$Symbol))
#[1] 21862
length(unique(xx$EntrezID))
#[1] 22606
## a few Symbols have additional EntrezIDs - probably psuedogenes that will be dropped.

write.table(xx$EntrezID, file="~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/DAVID/age15_DAVID_gene_background_2022-08-24.txt", 
            sep=",", row.names=F, col.names=F, quote=F)


## RUN DAVID https://david.ncifcrf.gov/summary.jsp
# GO_FAT_BP results, summarize using annotation cluster and enrichment score
# use top GO term in the category 
df <-read.csv("/Users/alexlussier/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/DAVID/aim3_DAVID_clusters_summary_2022-08-24.csv",
              header=F, stringsAsFactors = F)[-4]
head(df)

df$cluster.no <- c(1:sum(df$V5=="R2"), 1:sum(df$V5=="FDR"))
df$enrich.score <- as.numeric(sub("Enrichment Score: ", "", df$V2))
cluster.name <- strsplit(df$V3, "~")
df$cluster.name <- unlist(lapply(cluster.name, function(x) x[2]))

df <- df[, c(3:7)]
df$cluster.x <- c(sum(df$V5=="R2"):1, sum(df$V5=="FDR"):1)

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}
df$cluster.name <- sapply(df$cluster.name, firstup)
df$enrich.score
head(df)

df$V5 <- factor(df$V5, levels = c("FDR","R2"))

ggplot(df, aes(x=cluster.x, y=enrich.score, fill =V5)) + 
  geom_bar(stat="identity", alpha=0.7, 
           position = position_dodge2(preserve = "single")) +
  geom_hline(yintercept = -log10(0.05), color="red", lty=2, lwd=.6) + 
  scale_x_discrete(limits=c(18:1, 8:1),
                   labels=rev(df$cluster.name)) +
  labs(x=NULL, y="Enrichment Score (-log(p))") + 
  coord_flip() + 
  scale_fill_manual(values=c("steelblue4", "steelblue1"), 
                    labels = c("FDR-significant sites",
                               "R2-threshold sites"), "") + 
  theme_bw() + 
  theme(panel.grid=element_blank(), 
        text = element_text(family = "Gill Sans", size = 16))



rm(xx, background, df, fdr.genes, r2.genes, firstup, cluster.name)
max(age15.fdr.R2$P.value)

```

*GOseq - nothing much, not included in manuscript*
```{r}
load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/aim3_annot_table_2022-08-19.Rdata")

##perform GOseq, correcting for # CpGs/gene:
library(goseq)
library(FDb.InfiniumMethylation.hg19)
library(plyr)
library(dplyr)
library(tidyr)


# count # CpGs/gene:
# load all CpGs
load(file="~/Dropbox (Partners HealthCare)/Results Yiwen/BiologicalPsychiatry-2019/data/featureData.Rdata")
# subset to autosomal loci:
fData <- droplevels(featureData[featureData$TargetID %in% age15.results.variable$CpG, ])
rm(featureData)
hm450 <- get450k()
probes <- hm450[unique(fData$TargetID)]
x <- getNearestTSS(probes)
length(unique(x$nearestGeneSymbol))
#[1] 22420

# get counts for each gene:
gene.counts <- x %>% group_by(nearestGeneSymbol) %>%
  dplyr::summarize(n=n())

cpgs.per.gene <- gene.counts$n
names(cpgs.per.gene) <- gene.counts$nearestGeneSymbol
rm(gene.counts)

#FDR genes
y <- rep(0, length(cpgs.per.gene))
names(y) <- names(cpgs.per.gene)
rm(x, fData, hm450.controls, hm450, probes, seqinfo.hg19)
y[names(y) %in% fdr.annot2$NearestGene[fdr.annot2$FDR <0.05]] <- 1
sum(y)
#[1] 39
length(y)
#[1] 22420

pwf=nullp(DEgenes=y, "hg19", "geneSymbol", bias.data=cpgs.per.gene)
head(pwf)

GO.wall=goseq(pwf, "hg19", "geneSymbol")
nrow(GO.wall)
#[1] 22182
nrow(GO.wall[GO.wall$over_represented_pvalue < 0.05, ])
#[1] 312
## IMPORTANT: these are uncorrected p-values.
min(GO.wall$over_represented_pvalue) #0.000673182
GO.wall[GO.wall$over_represented_pvalue<0.001,] 
GO.wall[GO.wall$under_represented_pvalue<0.05,]

# calc FDR-adjusted p-values:
GO.wall$qValues = p.adjust(GO.wall$over_represented_pvalue, method="BH")
nrow(GO.wall[GO.wall$qValues < 0.05, ])
#[1] 0
## so none survive correction.

## R2 genes
y <- rep(0, length(cpgs.per.gene))
names(y) <- names(cpgs.per.gene)
rm(x, fData, hm450.controls, hm450, probes, seqinfo.hg19)
y[names(y) %in% fdr.annot2$NearestGene] <- 1
sum(y)
#[1] 39
length(y)
#[1] 22420

pwf.R2=nullp(DEgenes=y, "hg19", "geneSymbol", bias.data=cpgs.per.gene)
head(pwf)

GO.wall.R2=goseq(pwf.R2, "hg19", "geneSymbol")
nrow(GO.wall.R2)
#[1] 22182
nrow(GO.wall.R2[GO.wall.R2$over_represented_pvalue < 0.05, ])
#[1] 425
## IMPORTANT: these are uncorrected p-values.
min(GO.wall.R2$over_represented_pvalue) #2.564292e-06
GO.wall.R2[GO.wall.R2$over_represented_pvalue<0.001,] 
GO.wall.R2[GO.wall.R2$under_represented_pvalue<0.05,]

# calc FDR-adjusted p-values:
GO.wall.R2$qValues = p.adjust(GO.wall.R2$over_represented_pvalue, method="BH")
nrow(GO.wall.R2[GO.wall.R2$qValues < 0.05, ])
#[1] 0
## so none survive correction.

rm(pwf, GO.wall, pwf.R2, GO.wall.R2, y, cpgs.per.gene)
```

*ermineR - did not run* 
```{r, include=F}

price.annot <- read.table("/Users/alexlussier/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/price_annotation_450k.txt", 
                          header=T, sep='\t')

annot <- price.annot[which(price.annot$ID %in% age15.results.variable$CpG),]
head(annot)
age15.results.variable$Gene <- annot$Closest_TSS_gene_name[match(age15.results.variable$CpG,
                                                                 annot$ID)]
age15.fdr$Gene <- annot$Closest_TSS_gene_name[match(age15.fdr$CpG, annot$ID)]

load("../featureData.Rdata")
age15.results.variable$Gene.2 <- strsplit2(
  featureData$UCSC_REFGENE_NAME[match(age15.results.variable$CpG, featureData$TargetID)],
                              ";")[,1]
head(age15.results.variable)

#human.basic <- getGemmaAnnot(chipName = "Generic_human_ncbiIds", 
#                               chipFile = "~/Dropbox (Partners #HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Generic_human_ncbiIds.txt",
#                             annotType = "noParents",overwrite = T)

human.basic.file <- data.table::fread("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Generic_human_ncbiIds.txt", header=T)
dim(human.basic.file) #59125

GO.term.450k.alspac.variable <- human.basic.file[which(human.basic.file$GeneSymbols %in% 
                                                  age15.results.variable$Gene.2),]

#GSR - Biological processes
for(i in unique(age15.results.variable$Adversity)){
  print(i)
  if(i == 'abuse'){GSR.BP.results <- data.frame()}
  
  dat <- age15.results.variable[which(age15.results.variable$Adversity == i),]
  
  scores <- data.frame(Gene = dat$Gene.2, 
                       pvalue = dat$P.value)
  head(scores)
  head(GO.term.450k.alspac.variable)
  a <- gsr(annotation = GO.term.450k.alspac.variable, #'Generic_human_ncbiIds',
      scores = scores,
      scoreColumn = 2,
      iterations = 10000,
      bigIsBetter = FALSE,
      logTrans = TRUE, aspects = "B", minClassSize=5, geneReplicates = "best")

  a$results$GSR.type <- "BP"
  a$results$Adversity <- i
  a$results$Timing <- "All"
  
  for(x in levels(dat$Timing)){
    if(x== levels(dat$Timing)[1]){ temp <- data.frame()}
    print(paste(i, x, sep = " - "))
    b <- gsr(annotation = GO.term.450k.alspac.variable, #'Generic_human_ncbiIds',
      scores = scores[which(dat$Timing ==x),],
      scoreColumn = 2,
      iterations = 10000,
      bigIsBetter = FALSE,
      logTrans = TRUE, aspects = "B", minClassSize=5, geneReplicates = "best")
    
    b$results$GSR.type <- "BP"
    b$results$Adversity <- i
    b$results$Timing <- x
    temp <- rbind(temp, b$results)
    rm(b)
  }
  
  GSR.BP.results <- rbind(GSR.BP.results, a$results, temp)
  rm(a, temp)
}


levels(factor(GSR.BP.results$Adversity))
save(GSR.BP.results, 
     file ="~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Adversity_prelim/ermineR/GSR.BP.results.variable.probes.2021-06-06.Rdata")

#load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim #3/Adversity_prelim/ermineR/GSR.BP.results.variable.probes.20200527.Rdata", verbose=T)

load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Adversity_prelim/ermineR/GSR.BP.results.variable.probes.2021-06-06.Rdata")

### PLOTING ### 
GSR.BP.results$Timing <- factor(GSR.BP.results$Timing, levels = c("All","very_early","early",
                                                               "middle","late","accumulation",
                                                               "recency"))
dim(GSR.BP.results[which(GSR.BP.results$CorrectedPvalue<0.05),]) #21
GSR.BP.sig <- GSR.BP.results[which(GSR.BP.results$CorrectedPvalue<0.05),]
GSR.BP.sig.2 <- GSR.BP.sig[which(GSR.BP.sig$CorrectedMFPvalue<0.05),]
dim(GSR.BP.sig.2) #not doing that

GSR.BP.plot <- GSR.BP.results[which(GSR.BP.results$Name %in% GSR.BP.sig$Name),]

dim(GSR.BP.plot)
colnames(GSR.BP.plot)

GSR.BP.sig.2 <- GSR.BP.sig[c(order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="abuse")]),
                           order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="r_faminstability")]),
                           order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="Fscore")]),
                           order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="nbhqual")]),
                           order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="oneadult")]),
                           order(GSR.BP.sig$Pval[which(GSR.BP.sig$Adversity =="parcruelty")])
                         ),]

GSR.BP.sig.2 <- GSR.BP.sig[order(GSR.BP.sig$Adversity, GSR.BP.sig$CorrectedPvalue),]

ggplot(GSR.BP.sig.2,
       aes(x =Name, y= -log10(CorrectedPvalue), col=Timing, fill=Timing))+
  geom_hline(yintercept = -log10(0.05), col='red', linetype=2)+
  geom_bar(stat='identity', width=0.2)+
  geom_point(size=3)+
  scale_x_discrete("GO term", limits=rev(GSR.BP.sig.2$Name), labels=rev(GSR.BP.sig.2$Name)) + 
  #geom_point(data= GSR.BP.plot[which(GSR.BP.plot$CorrectedPvalue<0.05),], inherit.aes = T,
  #           size=3, position=position_dodge())+
  #facet_wrap(~Adversity, nrow=2, scales ='free')+
  theme_classic()+
  scale_fill_manual(values=c('darkgrey', viridis(n=6)[1:6]))+
  scale_color_manual(values=c('darkgrey', viridis(n=6)[1:6]))+
  #scale_fill_viridis(discrete=T)+
  theme(axis.text.y = element_text(size = 10))+
  geom_vline(xintercept = sum(GSR.BP.sig$Adversity=='parcruelty')+0.5)+
  geom_vline(xintercept = sum(GSR.BP.sig$Adversity %in% c('parcruelty',"oneadult"))+0.5)+
  geom_vline(xintercept = sum(GSR.BP.sig$Adversity %in% c('parcruelty',"oneadult","nbhqual"))+0.5)+
  geom_vline(xintercept = sum(GSR.BP.sig$Adversity %in% c('parcruelty',"oneadult","nbhqual",
                                                          "Fscore"))+0.5)+
    geom_vline(xintercept = sum(GSR.BP.sig$Adversity %in% c('parcruelty',"oneadult","nbhqual",
                                                          "Fscore","faminstability"))+0.5)+
  ylim(0,10)+
  ylab("-log10(FDR)")+
  coord_flip()

ggplot(GSR.BP.sig.2, aes(x = 1, fill=Adversity))+
  geom_bar(stat='count', position = position_stack())+
  scale_fill_manual(values = viridis(option='plasma', n=7, end =0.9)[-4])

rm(annot, dat, featureData, GO.term.450k.alspac.variable, GSR.BP.plot, GSR.BP.results,
    GSR.BP.sig,  GSR.BP.sig.2, i, human.basic.file, price.annot, scores, x)
```

*Kandaswamy 2020 lookup*
```{r}
kand.1 <- read.csv("../../Aim 3/Annotations/kandaswamy2020/t0001-10.1080_15592294.2020.1853317.csv", 
                   header = T, skip=1)
head(kand.1)
kand.2 <- read.csv("../../Aim 3/Annotations/kandaswamy2020/t0002-10.1080_15592294.2020.1853317.csv",
                   header = T, skip=1)
head(kand.2)

which(kand.1$Probe %in% age15.fdr.R2$CpG) #none
which(kand.2$Probe %in% age15.fdr.R2$CpG) #none

rm(kand.1, kand.2)

```
No overlap with relaxed threshold hits

*Puberty loci lookup*
```{r}
#pubertal age associations
pub.1 <- gdata::read.xls("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Annotations/Puberty_loci.xls", sheet=2, skip=1)
head(pub.1)
sum(age15.fdr.R2$CpG %in% pub.1$cpg.ID)#0

test <- age15.results.variable[which(age15.results.variable$CpG %in% pub.1$cpg.ID),]
ggplot(test, aes(x= Adversity, fill = Timing))+
  geom_bar(stat="count", position = "dodge")+
  scale_fill_viridis(discrete=T)+
  theme_classic()

min(test$P.value)
ggplot(test, aes(x = P.value, col = Adversity))+
  geom_density()+
  theme_classic()

### Pubertal hormones
pub.2 <- gdata::read.xls("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Annotations/Puberty_loci.xls", sheet=5, skip=1)
head(pub.2)
sum(age15.fdr.R2$CpG %in% pub.2$Testosterone)#0
sum(age15.fdr.R2$CpG %in% pub.2$LH)#0
sum(age15.fdr.R2$CpG %in% pub.2$FSH)#0
sum(age15.fdr.R2$CpG %in% pub.2$AMH)#0
sum(age15.fdr.R2$CpG %in% pub.2$Inhibin.B)#0


#Genes?#
pub.3 <- gdata::read.xls("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Aim 3/Annotations/Puberty_loci.xls", sheet=8, skip=1)

load("~/Dropbox (Partners HealthCare)/Dunn_shared/ALSPAC/Epigenetic/Paper Draft - Adversity->DNAm age17/Pacruelty_rerun/Results/aim3_annot_table_2022-08-19.Rdata", verbose=T)
for(i in fdr.annot2$NearestGene){
  print(fdr.annot2$NearestGene[grep(i, pub.3$Genes.associated.to.region)])
}
#Nada
for(i in fdr.annot2$NearestGene){
  print(fdr.annot2$NearestGene[grep(i, pub.1$Gene.symbol)])
}
#el zilcho

rm(fdr.annot2, i, pub.1, pub.2, pub.3, test)
```
No overlap with relaxed threshold hits


